# Hierarchical lineage architecture of human and avian spinal cord revealed by single-cell genomic barcoding
Scripts for analysis in Boezio et al (2025)

This repository contains the code, data references, and notebooks used in the analyses for the paper:

Hierarchical lineage architecture of human and avian spinal cord revealed by single-cell genomic barcoding
Giulia L. M. Boezio*, Jasper R. L. Depotter, Tom J. R. Frith, Arthur Radley, Stephanie Strohbuecker, Ana Cunha, Michael Howell, James Briscoe

https://www.biorxiv.org/content/10.1101/2025.10.24.684328v1.full


## üîç Overview

The Github includes:
- Python notebooks, python scripts and RMarkdown notebooks for data processing and figure generation
- Links to data hosted on GEO
- A reproducible environment for Python code


The Github does not include:
- Raw fastq and preprocessed Seurat objects (will be ready to download from GSE312424 soon)
- LARRY clone output (which can be generated by nf-LARRY pipeline available at: https://github.com/FrancisCrickInstitute/LARRY
  
---

## üìÅ Repository structure
```
‚îú‚îÄ‚îÄ metadata/              # metadata files as inputs for "permutation_analysis.ipynb"
‚îú‚îÄ‚îÄ R_markdowns/         # One RMarkdown notebook per figure+supplementary
‚îú‚îÄ‚îÄ Jupyter_notebooks/   # scRNAseq preprocessing and permutation lineage analyses
‚îú‚îÄ‚îÄ Python_scripts/           # Python scripts for cESFW and permutation analyses
‚îú‚îÄ‚îÄ environments/               # Environment files for cESFW and scRNAseq preprocessing/permutation analyses
‚îî‚îÄ‚îÄ README.md          # This file
```

## Environments: 
- cesfw_environment.yml is used for CESFW_feature_selection_chicken_data.ipynb and CESFW_feature_selection_human_data.ipynb
- scvi_environment.yml is used for all the other Jupyter notebook.

## Metadata:
- larry_chicken.csv contains the info to run permutation_analysis.ipynb with chicken data
- larry_human.csv contains the info to run permutation_analysis.ipynb with human data

## Jupyter notebooks/python scripts execution order + explanation + INPUTs/OUTPUTs:
```
	1.	Preprocessing_chicken_data.ipynb OR Preprocessing_human_data.ipynb (preprocessing of single cell data: QC, clean up, preliminary clustering, excluding contaminating cell types) 
  INPUT: cellranger output (specifically: filtered_feature_bc_matrix) OUTPUT: preprocessing_larry_*.h5ad 
	2.	cesfw_chicken_data.py OR cesfw_human_data.py (extracts CESFW features from the data) 
  INPUT: preprocessing_larry_*.h5ad OUTPUT: ESSs.npy, Used_Features.npy and EPs.npy
	3.	CESFW_feature_selection_chicken_data.ipynb OR CESFW_feature_selection_human_data.ipynb (CESFW features are clustered and selected by the Jupyter notebooks scripts (CESF*.ipynb))
  INPUT: ESSs.npy, Used_Features.npy, EPs.npy and preprocessing_larry_*.h5ad OUTPUT: selected_cesfw_features_*_data.npy
	4.	Integration_chicken_data.ipynb OR Integration_human_data.ipynb (creates the final integrated h5ad files using SCVI and CESFW features)
  INPUT: preprocessing_larry_*.h5ad, selected_cesfw_features_*_data.npy AND *_LARRY_*_clone_output.csv (deposited on GEO and obtained through nf-LARRY -see above- ) OUTPUT: larry_*.csv (metadata), preprocessing_larry_*_clusters.h5ad (final annotated object, soon available on GEO), optional: .csv to conver into Seurat object
	5.	permutation_analysis.ipynb (that uses permutation.py) 
  INPUT: larry_*.csv (metadata), permutation.py OUTPUT: lineage correlation matrix plots
	6.	wave_analysis.ipynb (for temporal wave analyses only)
  INPUT: larry_*.csv (metadata), permutation.py OUTPUT: lineage correlation matrix plots
```
